Add observability pillars (#81)
microsoft/code-with-engineering-playbook: observability/readme.md

3. Measure operational performance of the system.
4. Identify and diagnose failures to get to the problem fast.

## Pillars of observability

### Logging

Discrete events with the goal of helping engineers identify problem area(s) during failures.

> - Pay attention to logging levels. Logging too much will increase costs and decrease application throughtput.

Reviewer: ```suggestion
- Pay attention to logging levels. Logging too much will increase costs and decrease application throughput.
```

- Ensure logging configuration can be modified without code changes. Ideally, make it changeable without application restarts.
- If available, take advantage of logging levels per category allowing granular logging configuration.
- Check for log levels before logging, thus avoiding allocations and string manipulation costs.

> - Ensure service versions are included in logs, thus having the capability of identifying problematic releases.

Reviewer 2: Simpler structure - "Ensure service versions are included in logs to be able to identify problematic releases."

> - Log a raised exception only once. In your handlers, only catch expected exceptions that you can handle gracefully (even with a specific return code). If you want to log and rethrow, leave it to the top level exception handler. Do the minimal amount of cleanup work needed then throw to maintain the original stack trace. Don’t log a warning or stack trace for expected exceptions (eg: proper expected 404, 403).

Reviewer: ```suggestion
- Log a raised exception only once. In your handlers, only catch expected exceptions that you can handle gracefully (even with a specific return code). If you want to log and rethrow, leave it to the top level exception handler. Do the minimal amount of cleanup work needed then throw to maintain the original stack trace. Don’t log a warning or stack trace for expected exceptions (eg: properly expected 404, 403 HTTP statuses).
```

> - Log a raised exception only once. In your handlers, only catch expected exceptions that you can handle gracefully, including those for which you would set a specific return code. If you’re only going to log and rethrow (even without clobbering the stack trace), let the top level exception handler get it. If you have cleanup work to do on a catch, do only what is needed and then throw; to maintain the original stack trace. Don’t log a warning or stack trace for expected exceptions. Proper, expected 404s and 403s are not warnings.

Reviewer 3: Great thoughts on logging exceptions. I think this is a good practice, in general for logging exceptions. I think the best place for this will, perhaps be at a more generic level. I'm unsure where exactly, but will reach back to you with suggestions on this

> - Log a raised exception only once. In your handlers, only catch expected exceptions that you can handle gracefully, including those for which you would set a specific return code. If you’re only going to log and rethrow (even without clobbering the stack trace), let the top level exception handler get it. If you have cleanup work to do on a catch, do only what is needed and then throw; to maintain the original stack trace. Don’t log a warning or stack trace for expected exceptions. Proper, expected 404s and 403s are not warnings.

Reviewer 2: Would simplify and try and make a little clearer, something like:
"Log a raised exception only once. In your handlers, only catch expected exceptions that you can handle gracefully (even with a specific return code). If you want to log and rethrow, leave it to the top level exception handler. Do the minimal amount of cleanup work needed then throw to maintain the original stack trace. Don’t log a warning or stack trace for expected exceptions (eg: proper expected 404, 403).

- Fine tune logging levels in production (>= warning for instance). During a new release the verbosity can be increased to facilitate bug identification.

> - If sampling is used prefer implementing at the service level as opposed to defining in the logging system. This way we can ensure that exceptions (and other important information) will be logged. Another benefit is the reduced amount of roundtrips.

Reviewer 3: nit: If sampling is used, prefer implementing...

> - If sampling is used, prefer implementing at the service level as opposed to defining in the logging system. This way we can ensure that exceptions (and other important information) will be logged. Another benefit is the reduced amount of roundtrips.

Reviewer 2: Little edits in tone: "If using sampling, implement this at the service level rather than defining it in the logging system. This ensures that exceptions (and other important information) is logged. An additional benefit is reduced number of roundtrips."

> - Exclude health checks and non-business driven requests, unless they fail.

Reviewer 2: Flip the sentence: "Only include failures from health checks and non-business driven requests."

> - Ensure a downstream system mal-function won't cause massive repetitive logs being stored.

Reviewer 3: nit: malfunction

> - Ensure a downstream system malfunction won't cause massive repetitive logs being stored.

Reviewer 2: Would remove the word massive as makes it a bit dramatic ;(

Author: That's a massive change :)

- Don't reinvent the wheel, use existing tools to collect and analyse the data.
- Ensure personal identifiable information policies and restrictions are followed.

#### Having problems identifying what to log?

**At application startup**:

> - Unrecoverable errors from startup

Reviewer 2: Add a "."

- Warnings if application still runnable, but not as expected (i.e. not providing blob connection string, thus resorting to local files. Another example is if there's a need to fail back to a secondary service or a known good state, because it didn’t get an answer from a primary dependency.
- Information about the service’s state at startup (build #, configs loaded, etc.)

**Per incoming request**:

> - Basic information for each incoming request: scrubbed url, any user/tenant/request dimensions, response code returned, request-to-response latency, payload size, record counts, etc. (whatever you need to learn something from the aggregate data)

Reviewer: ```suggestion
- Basic information for each incoming request: the url (scrubbed of any personally identifying data, a.k.a. PII), any user/tenant/request dimensions, response code returned, request-to-response latency, payload size, record counts, etc. (whatever you need to learn something from the aggregate data)
```

## Tracing

Produces the information required to observe series of correlated operations in a distributed system. Once collected they show the path, measurements and faults in a end to end transaction.

- Ensure that at least key business transactions are traced.
- Include in each trace necessary information to identify software releases (i.e. service name, version). This is important to correlate deployments and system degradation.
- Ensure dependencies are included in trace (databases, I/O).
- If costs are a concern use sampling, avoiding throwing away errors, unexpected behaviour and critical information.
- Don't reinvent the wheel, use existing tools to collect and analyse the data.
- Ensure personal identifiable information policies and restrictions are followed.

## Metrics

Metrics provide a near real-time indication of how the system is running. As opposed to logs and traces, the amount of data collected using metrics remains constant as the system load increases.

- Measurements that need to be alerted upon are good candidates for metrics: queue length, retries, latency, requests/sec, cpu, memory. These metrics can be use to automatically scale components.

> - Use metrics to collect business success measurements: visits, sold items, session time, games played, jobs done. Ask stakeholders what they need.

Reviewer 3: Use metrics to collect business success measurements. Example - visits, sold items, session time, games played, jobs done. Ask stakeholders what they need
